### ISSUE TITLE
Nikon D90 - wrong FileNumber and DirectoryNumber

### ISSUE THREAD

**Issue by robyquin** [Exiv2/exiv2#880]
I'm using exiv2 version 0.27.1 on Linux and Windows,
and this bug occurs on both OS

The bug concerns the bad interpretation of the tags FileNumber and DirectoryNumber (and maybe others) in NEF Files of my Nikon D90, but there is not in NEF files of my Nikon D7500. 

If I take the values from NikonD90, and I divide them by 256, I get the correct values.

Is it possible that it depends on the different byteorder?
**MM in D90** (wrong interpretation)
**II in D7500** (good interpretation)

**Comment by clanmills** [comment]
Thanks for bringing this to our attention.  Do you have test files for both D90 and D7500 and I will investigate.

**Comment by robyquin** [comment]
I send you two photos, each name explains from which camera they come.

(format filename: **%Y%m%d_%H%M%S_FileNumber_Model.ext**)

in file D90, I get FileNumber = 4864 (instead of 19)

[20180913_205446_0019_NIKON_D90.zip](https://github.com/Exiv2/exiv2/files/3193747/20180913_205446_0019_NIKON_D90.zip)
[20180922_225401_0040_NIKON_D7500.zip](https://github.com/Exiv2/exiv2/files/3193748/20180922_225401_0040_NIKON_D7500.zip)

**Comment by clanmills** [comment]
Thanks for the files and can reproduce your issue:
```
$ exiv2 --grep Number/i ~/Downloads/*.NEF 
...  NIKON_D90.NEF  Exif.Photo.FNumber                           Rational    1  F7.1
...  NIKON_D90.NEF  Exif.Nikon3.SerialNumber                     Ascii       8  6900612
...  NIKON_D90.NEF  Exif.NikonLd3.LensIDNumber                   Byte        1  Nikon AF-S DX Nikkor 55-300mm f/4.5-5.6G ED VR
...  NIKON_D90.NEF  Exif.NikonFi.DirectoryNumber                 Short       1  25600 <---- ?
...  NIKON_D90.NEF  Exif.NikonFi.FileNumber                      Short       1  4864  <---- ?
...NIKON_D7500.NEF  Exif.Photo.FNumber                           Rational    1  F5.6
...NIKON_D7500.NEF  Exif.Nikon3.SerialNumber                     Ascii       8  6031721
...NIKON_D7500.NEF  Exif.NikonLd3.LensIDNumber                   Byte        1  Nikon AF-S DX Nikkor 55-300mm f/4.5-5.6G ED VR
...NIKON_D7500.NEF  Exif.NikonFi.DirectoryNumber                 Short       1  101
...NIKON_D7500.NEF  Exif.NikonFi.FileNumber                      Short       1  40
$ 
```

You are correct that the _D90 file is `MM` encoded and 7500 is `II`:
```
$ for i in ~/Downloads/*.NEF ; do exiv2 -pR $i|head -1; done
STRUCTURE OF TIFF FILE (MM): /Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF
STRUCTURE OF TIFF FILE (II): /Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF
$ file ~/Downloads/*.NEF
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF:   TIFF image data, big-endian, direntries=25, height=0, bps=0, compression=none, PhotometricIntepretation=RGB, manufacturer=NIKON CORPORATION, model=NIKON D90, orientation=lower-left, width=0
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF: TIFF image data, little-endian, direntries=28, height=120, bps=352, compression=none, PhotometricIntepretation=RGB, manufacturer=NIKON CORPORATION, model=NIKON D7500, orientation=upper-left, width=160
$ 
```

The Nikon3 "FileInfo" tag is 00x00b8.  You'll find the documentation here: https://exiv2.org/tags-nikon.html.  Both FileNumber and DirectoryNumber are "Short".  The binary decoder seems to be unaware that the D90 file is MM encoded.

```
$ grep -H nikonFiId src/nikonmn_int.cpp 
src/nikonmn_int.cpp:        TagInfo( 0, "Version", N_("Version"), N_("Version"), nikonFiId, makerTags, undefined, 4, printExifVersion),
src/nikonmn_int.cpp:        TagInfo( 6, "DirectoryNumber", N_("Directory Number"), N_("Directory number"), nikonFiId, makerTags, unsignedShort, 1, printValue),
src/nikonmn_int.cpp:        TagInfo( 8, "FileNumber", N_("File Number"), N_("File number"), nikonFiId, makerTags, unsignedShort, 1, printValue),
src/nikonmn_int.cpp:        TagInfo(0xffff, "(UnknownNikonFiTag)", "(UnknownNikonFiTag)", N_("Unknown Nikon File Info Tag"), nikonFiId, makerTags, unsignedByte, 1, printValue)
$
```

I'll debug the code and get this fixed for v0.27.2 which is scheduled to be released on 2019-06-30.

**Comment by robyquin** [comment]
For similitude with **canon array**, I would suggest replacing **littleEndian/bigEndian** with **invalidByteOrder**
(this change should also be verified on the other arrays)

in fact replacing it as follows and compiling on LINUX, it seems to work for me
 
in src/tiffimage_int.cpp
```cpp
//! Nikon File Info binary array - configuration
extern const ArrayCfg nikonFiCfg = {
        nikonFiId,        // Group for the elements
        littleEndian,     // Byte order <------------------------- invalidByteOrder // Use byte order from parent
        ttUndefined,      // Type for array entry
        notEncrypted,     // Not encrypted
        false,            // No size element
        true,             // Write all tags
        true,             // Concatenate gaps
        { 0, ttUnsignedByte,  1 }
    };
```

**Comment by clanmills** [comment]
Thanks.  I agree.  For FileNumber and DirectoryName the patch is:
```
586 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ git diff
diff --git a/src/tiffimage_int.cpp b/src/tiffimage_int.cpp
index 99986145..da9629fe 100644
--- a/src/tiffimage_int.cpp
+++ b/src/tiffimage_int.cpp
@@ -270,7 +270,7 @@ namespace Exiv2 {
     //! Nikon File Info binary array - configuration
     extern const ArrayCfg nikonFiCfg = {
         nikonFiId,        // Group for the elements
-        littleEndian,     // Byte order
+        invalidByteOrder, // Use byte order from parent
         ttUndefined,      // Type for array entry
         notEncrypted,     // Not encrypted
         false,            // No size element
587 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```
And it passes the test suite.

I didn't write this code, so although I know in general how it works, I'm not an expert on the makernotes being decoded.  I see that ArrayCfg's are defined with a mix of "inherit" and "little endian".   I see that all the "littleEndian" instances are nikon__Cfg.  My inclination is to change all to "inherit", however that causes a couple of failures in our test suite.

I'll investigate the test failures (by looking at the output of ExifTool) and decide in the next day or two what to do.

```
589 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ grep -e 'extern const ArrayCfg' -e littleEndian -e invalidByteOrder ../src/tiffimage_int.cpp 
    extern const ArrayCfg canonCsCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonSiCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonPaCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonCfCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonPiCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonTiCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonFiCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg canonPrCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonVrCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonPcCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonWtCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonAfCfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonAf2Cfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonAFTCfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonFiCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonMeCfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonFl1Cfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonFl2Cfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonFl3Cfg = {
        littleEndian,     // Byte order
    extern const ArrayCfg nikonLd1Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonLd2Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonLd3Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb1Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb2Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb2aCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb2bCfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb3Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg nikonCb4Cfg = {
        invalidByteOrder, // Use byte order from parent
    extern const ArrayCfg samsungPwCfg = {
        invalidByteOrder, // Use byte order from parent
590 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

Clearly, between us, we've found a fix for the "FileNumber" and "DirectoryNumber" issue.  And in the next couple of days, I will investigate if other ArrayCfg instances in the Nikon MakerNotes require similar treatment.

```
build/bin/exiv2 --grep Number ~/Downloads/*.NEF
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF  Exif.Photo.FNumber                           Rational    1  F7.1
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF  Exif.Nikon3.SerialNumber                     Ascii       8  6900612
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF  Exif.NikonLd3.LensIDNumber                   Byte        1  Nikon AF-S DX Nikkor 55-300mm f/4.5-5.6G ED VR
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF  Exif.NikonFi.DirectoryNumber                 Short       1  100
/Users/rmills/Downloads/20180913_205446_0019_NIKON_D90.NEF  Exif.NikonFi.FileNumber                      Short       1  19
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF  Exif.Photo.FNumber                           Rational    1  F5.6
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF  Exif.Nikon3.SerialNumber                     Ascii       8  6031721
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF  Exif.NikonLd3.LensIDNumber                   Byte        1  Nikon AF-S DX Nikkor 55-300mm f/4.5-5.6G ED VR
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF  Exif.NikonFi.DirectoryNumber                 Short       1  101
/Users/rmills/Downloads/20180922_225401_0040_NIKON_D7500.NEF  Exif.NikonFi.FileNumber                      Short       1  40
594 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

**Comment by clanmills** [comment]
There's no doubt that the correct fix is to use `invalidByteOrder` and not `littleEndian` for every instance of `ArrayCfg` in `src/nikonmn_int.cpp`. 

All the test failures involve the same metadatum `Exif.NikonAf.AFPointsInFocus` and file `test/data/exiv2-nikon-d70.jpg`.  The existing (incorrect) behaviour is:

```
824 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ exiv2 -pv --grep Points/i data/exiv2-nikon-d70.jpg 
0x0002 NikonAf      AFPointsInFocus             Short       1  256
825 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ exiv2 -pa --grep Points/i data/exiv2-nikon-d70.jpg 
Exif.NikonAf.AFPointsInFocus                 Short       1  Center
826 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ 
```

With the change:
```
826 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ ../build/bin/exiv2 -pv --grep Points/i data/exiv2-nikon-d70.jpg 
0x0002 NikonAf      AFPointsInFocus             Short       1  1
827 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ ../build/bin/exiv2 -pa --grep Points/i data/exiv2-nikon-d70.jpg 
Exif.NikonAf.AFPointsInFocus                 Short       1  Lower-right
828 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance/test $ 
```

When I submit the PR, I'll close this issue.  Appropriate changes to the test suite will be in the PR, so there's no need to add new tests.

**Comment by clanmills** [comment]
PR: https://github.com/Exiv2/exiv2/pull/885

**Comment by clanmills** [comment]
I put the following note in PR #885  I'll repeat it here as this may require additional work in future.  So better to record this twice to make it more likely to be discovered if required!

_I see there are instances of bigEndian in ArrayCfg in src/tiffimage_int.cpp When I changed them to invalidByteOrder, I got new test failures. The test bugfixes/redmine/test_issue_666.py writes a new Nikon MakerNote into an empty JPEG. Changing the default endian setting results in different byte order in the MakerNote. This is a non-trivial matter to investigate and determine if the current behaviour is incorrect. So, I will not touch the instances of bigEndian as no bug has been reported about writing Nikon MakerNotes._

_I know nothing about 666: https://dev.exiv2.org/issues/666._
